xbmc/windowing/wayland/WinSystemWayland.cpp

```diff
@@ -318,121 +363,276 @@ COutput* CWinSystemWayland::FindOutputByUserFriendlyName(const std::string& name
+
+    // Next buffer that the graphic context attaches will have the size corresponding
+    // to this configure, so go and ack it
+    AckConfigure(m_currentConfigureSerial);
+  }
+
+  bool wasInitialSetFullScreen = m_isInitialSetFullScreen;
+  m_isInitialSetFullScreen = false;
+
+  // Need to return true
+  // * when this SetFullScreen() call was initiated by a configure() event
+  // * on first SetFullScreen so GraphicsContext gets resolution
+  // Otherwise, Kodi must keep the old resolution.
+  return wasConfigure || wasInitialSetFullScreen;
+}
+
+
+void CWinSystemWayland::SetInhibitSkinReload(bool inhibit)
+{
+  m_inhibitSkinReload = inhibit;
+  if (!inhibit)
   {
-    m_shellSurface->SetFullScreen(output, m_fRefreshRate);
+    g_application.ReloadSkin();
   }
-  else
+}
+void CWinSystemWayland::HandleSurfaceConfigure(std::uint32_t serial, std::int32_t width, std::int32_t height)
+{
+  CSingleLock lock(g_graphicsContext);
+  CLog::LogF(LOGDEBUG, "Configure serial %u: size %dx%d", serial, width, height);
+  m_currentConfigureSerial = serial;
+  if (!ResetSurfaceSize(width, height, m_scale))
   {
-    // Shouldn't happen since we claim not to support windowed modes
-    CLog::Log(LOGWARNING, "Wayland windowing system asked to switch to windowed mode which is not really supported");
-    m_shellSurface->SetWindowed();
+    // nothing changed, ack immediately
+    AckConfigure(serial);
   }
+  // configure is acked when the Kodi surface has actually been reconfigured
+}
 
-  return true;
+void CWinSystemWayland::AckConfigure(std::uint32_t serial)
+{
+  // Send ack if we have a new serial number or this is the first time
+  // this function is called
+  if (serial != m_lastAckedSerial || !m_firstSerialAcked)
+  {
+    CLog::LogF(LOGDEBUG, "Acking serial %u", serial);
+    m_shellSurface->AckConfigure(serial);
+    m_lastAckedSerial = serial;
+    m_firstSerialAcked = true;
+  }
 }
 
-void CWinSystemWayland::HandleSurfaceConfigure(std::int32_t width, std::int32_t height)
+/**
+ * Set the internal surface size variables and perform resolution change
+ *
+ * Call only from Wayland event processing thread!
+ *
+ * \return Whether surface parameters changed and video resolution change was
+ *         performed
+ */
+bool CWinSystemWayland::ResetSurfaceSize(std::int32_t width, std::int32_t height, std::int32_t scale)
 {
   // Wayland will tell us here the size of the surface that was actually created,
   // which might be different from what we expected e.g. when fullscreening
   // on an output we chose - the compositor might have decided to use a different
   // output for example
   // It is very important that the EGL native module and the rendering system use the
   // Wayland-announced size for rendering or corrupted graphics output will result.
-  
-  CLog::Log(LOGINFO, "Got Wayland surface size %dx%d", width, height);
+
+  RESOLUTION switchToRes = RES_INVALID;
+
+  // FIXME See comment in SetFullScreen
+  CSingleLock lock(g_graphicsContext);
```

Comment:
You are right for `SetFullScreen` - `CGraphicContext` locks itself before calling it. The lock there is just for clarity to show that we're (ab)using `g_graphicsContext` for protecting our internal state. But in `ResetSurfaceSize` we need the lock since that function is called asynchronously on the Wayland event thread.